### PR TITLE
chore: send envelope (AR-912)

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We need to make the request sending the message envelope and handle the famous `412 ` response.
The current endpoint (`/conversations/{cnv}/otr/messages`) we support in `MessageApi` does not handle `QualifiedIds` and serializes the envelope as `JSON`.
The new endpoint (`/conversations/{cnv_domain}/{cnv}/proteus/messages`) only takes Protobuf and requires `QualifiedIds`.

### Solutions

* Added `sendEnvelope` to `MessageRepository`
* Added a new `qualifiedSendMessage` function to `MessageApi`.
* Added `Protobuf` to `network` in order to serialize these qualified protobuf envelopes.

### Dependencies

Needs releases with:

- [x] #232 
- [x] #231 

### Testing

Under construction.

### Notes

Some fields of the envelope are not yet supported for the scope of this PR, like `nativePush`, `ignore/report`, etc.

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApi.kt [hunk 1]
```diff
@@ -41,13 +58,38 @@ interface MessageApi {
...
+            val recipients: QualifiedUserToClientToEncMsgMap,
+            val nativePush: Boolean,
+            val priority: MessagePriority,
+            val transient: Boolean,
+            val `data`: String? = null,
+            val messageOption: QualifiedMessageOption
+        ) : Parameters()
     }
 
     suspend fun sendMessage(
```
Comment (MohamadJaara) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApi.kt:82]: @Deprecated ?

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApiImp.kt [hunk 2]
```diff
@@ -99,6 +106,37 @@ class MessageApiImp(private val httpClient: HttpClient) : MessageApi {
         }
     }
 
+    override suspend fun qualifiedSendMessage(
+        parameters: MessageApi.Parameters.QualifiedDefaultParameters,
+        conversationId: ConversationId
+    ): NetworkResponse<QualifiedSendMessageResponse> {
+        return try {
+            val response = httpClient.post("/conversations/${conversationId.domain}/${conversationId.value}/proteus/messages") {
```
Comment (MohamadJaara) [inline network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApiImp.kt:114]: extract the path strings to companion object ?

File: network/src/commonJvmAndroid/kotlin/com/wire/kalium/network/api/message/EnvelopeProtoMapperImpl.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
+package com.wire.kalium.network.api.message
+
+import com.wire.messages.Otr
+
+class EnvelopeProtoMapperImpl: EnvelopeProtoMapper {
+
+    private val otrClientEntryMapper = OtrClientEntryMapper()
+    private val otrUserIdMapper = OtrUserIdMapper()
+    private val otrClientIdMapper = OtrClientIdMapper()
```
[the remark this answers is not in the document]
Comment (vitorhugods) [reply network/src/commonJvmAndroid/kotlin/com/wire/kalium/network/api/message/EnvelopeProtoMapperImpl.kt:9]: They could, yes.
Not sure if they should.

In the long run, having the interaction between bigger and smaller mappers is annoying and costs too much time.

As no one else will use these `OtrUserIdMapper`, `OtrClientIdMapper`or `OtrClientEntryMapper`. I was thinking of considering `EnvelopeProtoMapper` an unit and testing from/to, comparying byteArrays and that's it.

**Review by MohamadJaara — COMMENTED** [review]
just some style and syntax suggestions, amazing work 😁

File: network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApi.kt (hunk 1 above)
Comment (vitorhugods) [reply network/src/commonMain/kotlin/com/wire/kalium/network/api/message/MessageApi.kt:82]: I guess it doesn't hurt :P 

I'll annotate.

**Pull request merged**